Update the spec to match current behavior of the zig compiler.

ehaas opened the pull request:
**Questions:**

1. Would you prefer this as 1 squashed commit or 1 PR per change instead of 1 PR with 9 commits? 
2. Some of the comments in the zig `parser.cpp` are out of date - should I make PRs for those as well?

**Changes:**

1. Add a script to compare zig parser to generated parser
2. Fix multidigit decimal integer parsing
3. Add support for sentinel terminators
4. Add support for optional errdefer payload
5. Fix binary and octal integer literal parsing
6. Add `anytype` support for struct fields
7. Allow missing end expression in SuffixOp
8. Add support for unicode character literals
9. Allow comments within multiline strings

SpexGuy commented on grammar/grammar.y:
```diff
@@ -332,21 +334,49 @@ ExprList <- (Expr COMMA)* Expr?
+dec_int <- dec dec_*
+hex_int <- hex hex_*
+
+utf8_trailer <- [�-�]
+oxf4 <- [�-�]
+oxf1_oxf3 <- [�-�]
+oxf0 <- [�-�]
+ox90_0xbf <- [�-�]
+oxee_oxef <- [�-�]
+oxed <- [�-�]
+ox80_ox9f <- [�-�]
+oxe1_oxec <- [�-�]
+oxe0 <- [�-�]
+oxa0_oxbf <- [�-�]
+oxc2_oxdf <- [�-�]
```
This is a good change, but is it possible to express these with escape sequences or hex literals so we don't need to require a special encoding for this file?

ehaas replied:
As far as I can tell peg/leg supports octal escapes but not hex. I replaced the latin1 chars with octal escapes so ASCII/utf8 is fine for this file again.

SpexGuy reviewed (changes requested):
This looks great!  Thanks for taking the time to update this.  To answer your questions:

> Would you prefer this as 1 squashed commit or 1 PR per change instead of 1 PR with 9 commits?

Yes, I'd appreciate if you would squash the commits.

> Some of the comments in the zig parser.cpp are out of date - should I make PRs for those as well?

Please do!  PRs that improve documentation are always welcome.

---

The test script is a fantastic idea.  If you have tests you've been using, I'd love to pull those into the repo also, maybe under `grammar/tests/`.

ehaas commented:
I squashed the commits and added a `tests` dir that contains a bunch of zig files (all copied from `test/stage1/behavior` in the main zig repo). All the tests are passing on my machine.

One current issue is that this grammar does not understand doc comments (it just treats them like regular comments); I can look into adding those next.

SpexGuy commented:
Nice, that's a lot more test than I expected!  Thanks for updating the characters.  I'll merge this in.
